be/src/formats/csv/converter.cpp

```diff
@@ -75,4 +75,57 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
     return nullable ? std::make_unique<NullableConverter>(std::move(c)) : std::move(c);
 }
 
+// Hive collection delimiter generate rule refer to:
+// https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L250
+static char get_collection_delimiter(char mapkey_delimiter, size_t nested_array_level) {
+    CHECK(nested_array_level > 1 && nested_array_level < 153);
+
+    if (nested_array_level == 2) {
+        return mapkey_delimiter;
+    }
+
+    // tmp maybe negative, dont use size_t.
+    int32_t tmp;
+    if (nested_array_level <= 7) {
+        // [3, 7] -> [4, 8]
+        tmp = static_cast<int32_t>(nested_array_level) + (4 - 3);
+    } else if (nested_array_level <= 20) {
+        // [8, 20] -> [14, 26]
+        tmp = static_cast<int32_t>(nested_array_level) + (14 - 8);
+    } else if (nested_array_level <= 24) {
+        // [21, 24] -> [28, 31]
+        tmp = static_cast<int32_t>(nested_array_level) + (28 - 21);
+    } else if (nested_array_level <= 152) {
+        // [25, 152] -> [-128, -1]
+        tmp = static_cast<int32_t>(nested_array_level) + (-128 - 25);
+    }
+
+    return (char)tmp;
```

Comment:
static_cast

Reply:
[DorianZheng](https://github.com/DorianZheng) already mentioned it. 🤣